### PR TITLE
Add FourierPositionalEmbedding to physicsnemo.nn

### PR DESCRIPTION
**Pull request by mnabian**
# PhysicsNeMo Pull Request

## Description

Adds a deterministic, axis-wise (NeRF-style) Fourier positional embedding for continuous coordinates as a reusable `physicsnemo.nn` layer. It fills a gap in the existing Fourier/positional family: `FourierEmbedding` uses random Gaussian frequencies on a scalar, and `FourierLayer` is a learned/projected (axis-mixing) map — neither provides the standard deterministic per-axis encoding used by implicit/INR-style decoders.

What it does
- Encodes each coordinate axis with a fixed frequency schedule (geometric/octave by default, or an explicit `freqs` tensor) into sin/cos features, optionally concatenating the raw input.
- No learnable parameters; frequencies stored as a non-persistent buffer. Exposes `out_dim` for sizing downstream layers and validates the input feature dimension.

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: physicsnemo/nn/module/embedding_layers.py [hunk 1]
```diff
@@ -257,3 +258,125 @@ def forward(self, t: Float[Tensor, "B ..."]) -> Float[Tensor, "B D"]:
...
+        r"""Output feature dimension :math:`D_{out}`."""
+        base = self.in_dim if self.include_input else 0
+        return base + 2 * self.in_dim * self.num_bands
+
+    def forward(
+        self, x: Float[Tensor, "... in_dim"]
+    ) -> Float[Tensor, "... out_dim"]:
+        r"""Encode coordinates ``x``; see the class docstring for shapes."""
+        if x.shape[-1] != self.in_dim:
+            raise ValueError(f"Expected last dim {self.in_dim}, got {x.shape[-1]}")
```
[the remark this answers is not in the document]
Comment (mnabian) [reply physicsnemo/nn/module/embedding_layers.py]: Fixed.

File: physicsnemo/nn/module/embedding_layers.py [hunk 2]
```diff
@@ -257,3 +258,125 @@ def forward(self, t: Float[Tensor, "B ..."]) -> Float[Tensor, "B D"]:
         r"""Forward pass computing soft one-hot embeddings."""
         ind = (t.view(-1, 1) * (self.num_channels - 1)).to(self.indices.dtype)
         return torch.clamp(1 - torch.abs(ind - self.indices), min=0)
+
+
+class FourierPositionalEmbedding(torch.nn.Module):
```
[the remark this answers is not in the document]
Comment (CharlelieLrt) [reply physicsnemo/nn/module/embedding_layers.py]: That's true, but this one looks very safe to put directly in production. No need to put to experimental

[the remark this answers is not in the document]
Comment (mnabian) [reply physicsnemo/nn/module/embedding_layers.py]: Fixed

File: test/nn/module/test_embedding_layers.py [hunk 3]
```diff
@@ -405,3 +407,120 @@ def test_deterministic(self, device):
...
+    with pytest.raises(ValueError):
+        FourierPositionalEmbedding(in_dim=0)
+    with pytest.raises(ValueError):
+        FourierPositionalEmbedding(in_dim=3, num_bands=0)
+
+
+def test_fourier_positional_embedding_state_dict_no_freqs(device):
+    # freqs is a non-persistent buffer; it must not appear in the state dict.
+    emb = FourierPositionalEmbedding(in_dim=3, num_bands=4).to(device)
+    assert "freqs" not in emb.state_dict()
```
[the remark this answers is not in the document]
Comment (mnabian) [reply test/nn/module/test_embedding_layers.py]: Fixed.

File: physicsnemo/nn/module/embedding_layers.py [hunk 4]
```diff
@@ -257,3 +258,128 @@ def forward(self, t: Float[Tensor, "B ..."]) -> Float[Tensor, "B D"]:
...
+    num_bands : int, optional, default=10
+        Number of frequency bands when ``freqs`` is not given.
+    include_input : bool, optional, default=True
+        Prepend the raw input coordinates to the embedding.
+    base_freq : float, optional, default=math.pi
+        Frequency of the first band for the generated schedule.
+    freq_scale : float, optional, default=2.0
+        Geometric ratio between consecutive band frequencies for the
+        generated schedule.
+    freqs : torch.Tensor or sequence of float, optional, default=None
```
Comment (CharlelieLrt) [inline physicsnemo/nn/module/embedding_layers.py]: nit: missing shape for tensor arg, e.g. " tensor of shape :math:`(B, C, H, W)`."

Edit: I might have gotten confused, because here it's written 1D (in which case indicating the shape here is indeed useless), but later in the code it's using a `flatten` so it might accept non-1D freqs?

File: physicsnemo/nn/module/embedding_layers.py [hunk 5]
```diff
@@ -257,3 +258,128 @@ def forward(self, t: Float[Tensor, "B ..."]) -> Float[Tensor, "B D"]:
...
+    """
+
+    def __init__(
+        self,
+        in_dim: int = 3,
+        num_bands: int = 10,
+        include_input: bool = True,
+        base_freq: float = math.pi,
+        freq_scale: float = 2.0,
+        freqs: Tensor | None = None,
```
Comment (CharlelieLrt) [inline physicsnemo/nn/module/embedding_layers.py]: nit: add jaxtyping shape annotation for tensor argument

File: physicsnemo/nn/module/embedding_layers.py [hunk 6]
```diff
@@ -257,3 +258,128 @@ def forward(self, t: Float[Tensor, "B ..."]) -> Float[Tensor, "B D"]:
...
+        r"""Number of frequency bands in the schedule."""
+        return int(self.freqs.numel())
+
+    @property
+    def out_dim(self) -> int:
+        r"""Output feature dimension :math:`D_{out}`."""
+        base = self.in_dim if self.include_input else 0
+        return base + 2 * self.in_dim * self.num_bands
+
+    def forward(self, x: Float[Tensor, "... in_dim"]) -> Float[Tensor, "... out_dim"]:
```
Comment (CharlelieLrt) [inline physicsnemo/nn/module/embedding_layers.py]: x: Float[Tensor, "... in_dim"] -> Float[Tensor, "... out_dim"] means that the leading dimensions of input and output do not need to match. Is that the intended behavior?
Otherwise maybe `x: Float[Tensor, "*dims in_dim"] -> Float[Tensor, "*dims out_dim"]` is more appropriate?

File: test/nn/module/test_embedding_layers.py [hunk 7]
```diff
@@ -405,3 +407,145 @@ def test_deterministic(self, device):
...
+
+
+def test_fourier_positional_embedding_out_dim_and_shape(device):
+    emb = FourierPositionalEmbedding(in_dim=3, num_bands=4).to(device)
+    assert emb.num_bands == 4
+    assert emb.out_dim == 3 + 2 * 3 * 4  # 27
+    y = emb(torch.randn(5, 3, device=device))
+    assert y.shape == (5, 27)
+    # No learnable parameters.
+    assert sum(p.numel() for p in emb.parameters()) == 0
```
Comment (CharlelieLrt) [inline test/nn/module/test_embedding_layers.py:446]: This test is not necessary (test that only check the output shape are to be avoided because too permissive). The test pattern should be:
- 1 test function for the constructor (test both default and non-default constructor args)
- 1 test function for non-regression testing of the forward. This should test not only the shape of the output from the forward, but also the values against reference data (either known reference output or golden files)

File: test/nn/module/test_embedding_layers.py [hunk 8]
```diff
@@ -405,3 +407,145 @@ def test_deterministic(self, device):
...
+    )
+    assert emb.out_dim == 2 * 2 * 3  # 12
+    assert emb(torch.zeros(4, 2, device=device)).shape == (4, 12)
+
+
+def test_fourier_positional_embedding_leading_dims(device):
+    emb = FourierPositionalEmbedding(in_dim=3, num_bands=2).to(device)
+    for shape in [(3,), (7, 3), (2, 7, 3)]:
+        out = emb(torch.randn(*shape, device=device))
+        assert out.shape == (*shape[:-1], emb.out_dim)
```
Comment (CharlelieLrt) [inline test/nn/module/test_embedding_layers.py]: Same comment here. The tesing pattern can be simplified and streamlined

File: test/nn/module/test_embedding_layers.py [hunk 9]
```diff
@@ -405,3 +407,145 @@ def test_deterministic(self, device):
...
+                math.cos(2.0 * 0.3),
+                math.sin(1.0 * 0.7),
+                math.sin(2.0 * 0.7),
+                math.cos(1.0 * 0.7),
+                math.cos(2.0 * 0.7),
+            ]
+        ],
+        device=device,
+    )
+    torch.testing.assert_close(out, expected)
```
Comment (CharlelieLrt) [inline test/nn/module/test_embedding_layers.py]: Consider using pytest parameterization instead of writing a separate function for each config

**Review by CharlelieLrt — CHANGES_REQUESTED** [review]
Just some details here and there, nothing major

File: physicsnemo/nn/module/embedding_layers.py (hunk 4 above)
Comment (mnabian) [reply physicsnemo/nn/module/embedding_layers.py]: Fixed here: 34e6682c779eea8d5ba868fcf0052a2f4c8ebd71

File: physicsnemo/nn/module/embedding_layers.py (hunk 5 above)
Comment (mnabian) [reply physicsnemo/nn/module/embedding_layers.py]: Fixed here: 34e6682c779eea8d5ba868fcf0052a2f4c8ebd71

File: physicsnemo/nn/module/embedding_layers.py (hunk 6 above)
Comment (mnabian) [reply physicsnemo/nn/module/embedding_layers.py]: Good catch! Fixed here: 34e6682c779eea8d5ba868fcf0052a2f4c8ebd71

File: test/nn/module/test_embedding_layers.py (hunk 7 above)
Comment (mnabian) [reply test/nn/module/test_embedding_layers.py:446]: Fixed here: 34e6682c779eea8d5ba868fcf0052a2f4c8ebd71

File: test/nn/module/test_embedding_layers.py (hunk 8 above)
Comment (mnabian) [reply test/nn/module/test_embedding_layers.py]: Fixed here: 34e6682c779eea8d5ba868fcf0052a2f4c8ebd71

File: test/nn/module/test_embedding_layers.py (hunk 9 above)
Comment (mnabian) [reply test/nn/module/test_embedding_layers.py]: Fixed here: 34e6682c779eea8d5ba868fcf0052a2f4c8ebd71

**Review by CharlelieLrt — APPROVED** [review]
LGTM!

**Review by loliverhennigh — APPROVED** [review]
LGTM
